tests/overlay_tests.cpp

```diff
@@ -1343,6 +1340,75 @@ TEST_F(OverlayTest, ROOT_checkAddVirtualNetworks)
   ASSERT_EQ(agentOverlay->info().subnet6(), "fd04::/64");
 }
 
+
+// Tests the ability of the `Agent overlay module` to be registered
+// when master failover happens
+TEST_F(OverlayTest, checkAgentResetConfigurationAttempts)
+{
+  Try<Owned<cluster::Master>> master = StartMaster();
+  ASSERT_SOME(master);
+  StandaloneMasterDetector *detector =
+    new StandaloneMasterDetector(master.get()->pid);
+
+  LOG(INFO) << "Master PID: " << master.get()->pid;
+
+  // Ask overlay Master to use the replicated log by setting
+  // `replicated_log_dir`. We are not specifying `zk` configuration so
+  // the `quorum` will default to "1".
+  MasterConfig masterOverlayConfig;
+  masterOverlayConfig
+    .set_replicated_log_dir("overlay_replicated_log");
+
+  Try<Owned<Anonymous>> masterModule = startOverlayMaster(masterOverlayConfig);
+  ASSERT_SOME(masterModule);
+
+  // Master `Anonymous` module created successfully.
+  UPID overlayMaster = UPID(
+      MASTER_MANAGER_PROCESS_ID,
+      master.get()->pid.address);
+
+  AgentConfig agentOverlayConfig;
+  agentOverlayConfig.set_master(stringify(overlayMaster.address));
+  // Set number of times the agent will attempt to configure virtual
+  // networks by re-registering with the master.
+  agentOverlayConfig.set_max_configuration_attempts(1);
+
+  // Setup futures to notify the test that Agent overlay module has
+  // registered.
+  Future<AgentRegisteredAcknowledgement> agentRegisteredAcknowledgement =
+    FUTURE_PROTOBUF(AgentRegisteredAcknowledgement(), _, _);
+
+  Try<Owned<overlayAgent::ManagerProcess>> agentModule = startOverlayAgent(
+      agentOverlayConfig, detector);
+
+  AWAIT_READY(agentRegisteredAcknowledgement);
+
+  ASSERT_SOME(agentModule);
+
+  // Check the agent is allowed to progress.
+  AWAIT_READY(agentModule.get()->ready());
+
+  // Simulate a new master detected event to the agent module
+  // StandaloneMasterDetector detector(master.get()->pid);
+  detector->appoint(master.get()->pid);
+
+  // Wait for the agent module to (re-)register.
+  agentRegisteredAcknowledgement = FUTURE_PROTOBUF(
+    AgentRegisteredAcknowledgement(), _, _);
+  AWAIT_READY(agentRegisteredAcknowledgement);
```

Comment:
@urbanserj shouldn't we be exposing the config attempts and checking the number of config attempts?

Reply:
@asridharan Could you please check the updated version?

https://github.com/dcos/dcos-mesos-modules/blob/d3399b25db3dd2961506a483715695b4559f1ba6/tests/overlay_tests.cpp#L1402-L1421